com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/eval/JdtEvaluationProvider.java

```diff
@@ -0,0 +1,271 @@
+                        logger.severe(String.format("Cannot initialize project: %s", e.toString()));
+                    }
+                }
+            }
+
+            if (project == null) {
+                completableFuture.completeExceptionally(new IllegalStateException(String.format("Project %s cannot be found.", projectName)));
+                return completableFuture;
+            }
+            if (launch == null) {
+                launch = createLaunch(project);
+            }
+        }
+
+
+        ThreadReference thread = sf.thread();
+        if (debugTarget == null) {
+            debugTarget = new JDIDebugTarget(launch, thread.virtualMachine(), "", false, false, null, false) {
+                @Override
+                protected synchronized void initialize() {
+                    // use empty initialize intentionally to avoid to register jdi event listener
+                }
+            };
+        }
+        try {
+            JDIThread jdiThread = getJDIThread(thread);
+
+            synchronized (jdiThread) {
+                if (jdiThread.isPerformingEvaluation()) {
+                    jdiThread.wait();
+                }
+
+                ASTEvaluationEngine engine = new ASTEvaluationEngine(project, debugTarget);
+                JDIStackFrame stackframe = createStackFrame(sf);
+
+                ICompiledExpression ie = engine.getCompiledExpression(code, stackframe);
+                engine.evaluateExpression(ie, stackframe, evaluateResult -> {
+                    synchronized (jdiThread) {
+                        jdiThread.notify();
+                    }
+                    if (evaluateResult == null || evaluateResult.hasErrors()) {
+                        Exception ex = evaluateResult.getException() != null ? evaluateResult.getException()
+                                : new RuntimeException(StringUtils.join(evaluateResult.getErrorMessages()));
+                        completableFuture.completeExceptionally(ex);
+                    } else {
+                        try {
+                            // we need to read fValue from the result Value instance implements by JDT
+                            Value value = (Value) FieldUtils.readField(evaluateResult.getValue(), "fValue", true);
+                            completableFuture.complete(value);
+                        } catch (IllegalArgumentException | IllegalAccessException ex) {
+                            completableFuture.completeExceptionally(ex);
+                        }
+                    }
+                }, 0, false);
+            }
+
+        } catch (Exception ex) {
+            completableFuture.completeExceptionally(ex);
+        }
+        return completableFuture;
+    }
+
+    private JDIStackFrame createStackFrame(StackFrame sf) {
+        return new JDIStackFrame(getJDIThread(sf.thread()), sf, 0);
+    }
+
+    private JDIThread getJDIThread(ThreadReference thread) {
+        synchronized (threadMap) {
```

Comment:
No need for a map. The JDIThread is just a thing wrap that we can create on the fly.

Reply:
No, the eval task is referenced in JDIThread instance, so we need to save the create JDIThread instance